internal/controller/groupqualityofservicelimits/groupqualityofservicelimits.go

```diff
@@ -198,11 +247,75 @@ func (c *external) Delete(ctx context.Context, mg resource.Managed) (managed.Ext
 		return managed.ExternalDelete{}, errors.New(errNotGroupQualityOfServiceLimits)
 	}
 
-	fmt.Printf("Deleting: %+v", cr)
+	cr.SetConditions(xpv1.Deleting())
+
+	user := cloudian.User{
+		GroupID: cr.Spec.ForProvider.GroupID,
+		UserID:  "*",
+	}
+	err := c.cloudianService.DeleteQOS(ctx, user, cr.Spec.ForProvider.Region)
+	if err != nil && !errors.Is(err, cloudian.ErrNotFound) {
+		return managed.ExternalDelete{}, errors.Wrap(err, errGetCreds)
+	}
 
 	return managed.ExternalDelete{}, nil
 }
 
 func (c *external) Disconnect(ctx context.Context) error {
 	return nil
 }
+
+func toCloudianQos(qos v1alpha1.GroupQualityOfServiceLimitsParameters) (cloudian.QualityOfService, error) {
+	warning, err := toCloudianLimits(qos.Warning)
+	if err != nil {
+		return cloudian.QualityOfService{}, err
+	}
+	high, err := toCloudianLimits(qos.Hard)
+	if err != nil {
+		return cloudian.QualityOfService{}, err
+	}
+	return cloudian.QualityOfService{
+		Warning: warning,
+		Hard:    high,
+	}, nil
+}
+
+func toCloudianLimits(limits *v1alpha1.QualityOfServiceLimits) (cloudian.QualityOfServiceLimits, error) {
+	if limits == nil {
+		return cloudian.QualityOfServiceLimits{}, nil
+	}
+
+	var err error
+	qosl := cloudian.QualityOfServiceLimits{}
+
+	if qosl.StorageQuotaKiBs, err = limits.StorageQuotaBytes.ToKiB(); err != nil {
+		return cloudian.QualityOfServiceLimits{}, err
+	}
+	if qosl.InboundKiBsPerMin, err = limits.InboundBytesPerMin.ToKiB(); err != nil {
+		return cloudian.QualityOfServiceLimits{}, err
+	}
+	if qosl.OutboundKiBsPerMin, err = limits.OutboundBytesPerMin.ToKiB(); err != nil {
+		return cloudian.QualityOfServiceLimits{}, err
+	}
+
+	if limits.StorageQuotaCount != nil {
+		qosl.StorageQuotaCount = ptr.To(int64(*limits.StorageQuotaCount))
+	}
+	if limits.RequestsPerMin != nil {
+		qosl.RequestsPerMin = ptr.To(int64(*limits.RequestsPerMin))
+	}
+
+	return qosl, nil
+}
+
+func limitsEqual(a cloudian.QualityOfServiceLimits, b cloudian.QualityOfServiceLimits) bool {
+	eq := func(a *int64, b *int64) bool {
+		return (a == b) || (a != nil && b != nil && *a == *b)
+	}
+
+	return eq(a.InboundKiBsPerMin, b.InboundKiBsPerMin) &&
```

Comment:
```suggestion
	return ptr.Equal(a.InboundKiBsPerMin, b.InboundKiBsPerMin) &&
```

And delete the function variable. Etc....